astro_reduction_tool.py

```diff
@@ -0,0 +1,222 @@
+        raise ValueError(f"No directory {data_path}")
+    if len(_ifc_mf) == 0:
+        ccd_collection = ccdp.ImageFileCollection(data_path, keywords=keys)
+        if _filter:
+            ccd_collection = ccd_collection.files_filtered(filter=_filter)
+        else:
+            ccd_collection = ccd_collection.files
+    else:
+        ccd_collection = _ifc_mf
+    if len(ccd_collection) != 0:
+        logger.info(f"Loading data from: {data_path}")
+        for ccd_path in ccd_collection:
+            dir = os.path.join(data_path, ccd_path)
+            ccd_file = CCDData(
+                fits.getdata(dir), meta=fits.getheader(dir, keywords=keys), unit="adu"
+            )
+            header_value = ccd_file.meta.get("combined")
+            if header_value is None or len(_ifc_mf) != 0:
+                data_list.append(ccd_file)
+                name_list.append(ccd_path)
+                now = datetime.now()
+                formatted_time = now.strftime("[%Y-%m-%dT%H:%M:%S.%f]")
+                logging.info(
+                    "%s  Loading %s file %s (1 of 1)",
+                    formatted_time,
+                    ccd_file.header["IMAGETYP"],
+                    ccd_path,
+                )
+    ccd_data_list = ccd_data(data_list,name_list)
+    return ccd_data_list
+
+def debias_dedark(file_list, master_dark, master_bias):
+    for h in range(0, len(file_list.data_list)):
+        ccd_file = file_list.data_list[h]
+        ccd_name = file_list.name_list[h]
+        bias_subtracted = ccdp.subtract_bias(ccd_file, master_bias)
+        file_list.data_list[h] = ccdp.subtract_dark(
+            bias_subtracted,
+            master_dark,
+            exposure_time="EXPTIME",
+            exposure_unit=u.adu,
+            scale=True,
+        )
+        now = datetime.now()
+        formatted_time = now.strftime("[%Y-%m-%dT%H:%M:%S.%f]")
+        logging.info(
+            "%s  Debias and dedark %s file %s (%d of %d)",
+            formatted_time,
+            ccd_file.header["IMAGETYP"],
+            ccd_name,
+            h+1,
+            len(file_list.data_list)
+        )
+    return file_list.data_list
+
+
+def master_flat_generator(flat_p, master_dark, master_bias, filt):
+    m_flat = []
+    m_flat_n = []
+    keys = ["imagetyp", "filter"]
+    for f in filt:
+        flat_data = read_path_into_CCDData(flat_p, keys, f)
+        flat_subtracted_list = debias_dedark(flat_data, master_dark, master_bias)
+        combined_flat = ccdp.combine(
+            flat_subtracted_list,
+            method="median",
+            unit=u.adu,
+            scale=inv_median,
+            sigma_clip=True,
+            sigma_clip_low_thresh=5.0,
+            sigma_clip_high_thresh=5.0,
+            sigma_clip_func=np.ma.median,
+            sigma_clip_dev_func=mad_std,
+            flat=2,
+        )
+        combined_flat.meta["combined"] = True
+        name = os.path.join(flat_p,"master_flat_{}.fits".format(f))
```

Comment:
Warto rozwazyc uzycie f-stringow:
```suggestion
        name = os.path.join(flat_p, f"master_flat_{f}.fits")
```